Library/OcMachoLib/Header.c

```diff
@@ -160,20 +179,89 @@ MachoFilterFatArchitecture64 (
 
       if (Offset == 0
         || OcOverflowAddU32 (Offset, Size, &TmpSize)
-        || TmpSize > *FileSize) {
+        || TmpSize > FullSize) {
         return FALSE;
       }
 
-      *FileData = *FileData + Offset;
-      *FileSize = Size;
-
+      *FatOffset  = Offset;
+      *FatSize    = Size;
       return TRUE;
     }
   }
 
   return FALSE;
 }
 
+/**
+  Moves file pointer and size to point to specified slice in case
+  FAT Mach-O is used.
+
+  @param[in,out] FileData  Pointer to pointer of the file's data.
+  @param[in,out] FileSize  Pointer to file size of FileData.
+  @param[in]     CpuType   Desired CPU slice to use.
+
+  @return FALSE is not valid FAT image.
+**/
+BOOLEAN
+MachoFilterFatArchitectureByType (
+  IN OUT UINT8         **FileData,
+  IN OUT UINT32        *FileSize,
+  IN     MACH_CPU_TYPE CpuType
+  )
+{
+  UINT32        FatOffset;
+  UINT32        FatSize;
+
+  ASSERT (FileData != NULL);
+  ASSERT (FileSize != NULL);
+  ASSERT (*FileSize > 0);
+
+  if (!MachoGetFatArchitectureOffset (*FileData, *FileSize, *FileSize, CpuType, &FatOffset, &FatSize)) {
+    return FALSE;
+  }
+
+  *FileData = FileData[FatOffset];
```

Comment:
That's not right, should be (*FileData) + FatOffset. Or maybe *FileData += FatOffset.